Add feature:  Ignore lists

push0ret opened the pull request:
Hi!

Thank you for all your work with mpd/mpdscribble. It's a seriously great project!

I'm hoping to contribute with this little feature I built for my own use.
It simply allows you to specify an ignore list per scrobbler in which you can specify tracks for the scrobbler to ignore.

I decided to make it per scrobbler, as for my personal use case, I have a file scrobbler that does not have excludes while last.fm has some excludes.

The PR includes a manpage update describing the file format in detail.
I'm adding the rendered excerpt here for your convenience.

```
IGNORE FILE FORMAT
       Tracks can be ignored by listing them in an ignore file.  Each line  in
       the  file  specifies a pattern to match tracks you wish to exclude from
       scrobbling.  The format is simple and flexible, allowing you  to  match
       by artist, album, title, track number or a combination of these fields.

   File Format
       A  tag match is specified as tagname="value", where tagname is replaced
       with one of the supported tags (artist, album, title,  track).   Values
       must be quoted and spaces are not allowed surrounding the equal sign.

       Each line consists of one or more tag matches separated by spaces:
           tag1="value1" tag2="value2" ...

       Tags can appear in any order and blank lines are ignored.

       A  backslash  (\) is interpreted as escape character and may be used to
       escape literal double quotes within a value.  To write a literal  back-
       slash, use a double backslash (\\).

       Each  line  is limited to 4096 characters including the newline charac-
       ter.  Superflous characters are silently ignored.


   Examples
       If a tag is omitted, any value for that field is matched:

       title="Bohemian Rhapsody"
              Matches any track titled Bohemian Rhapsody.

       artist="Queen"
              Matches any track by Queen, regardless of album or title.

       artist="Queen" album="A Night at the Opera"
              Matches any track by Queen from A Night at the Opera, regardless
              of title.

       artist="Queen" album="A Night at the Opera" title="Bohemian Rhapsody"
              Matches a specific track by Queen, from A Night  at  the  Opera,
              titled Bohemian Rhapsody.

       artist="Queen" album="A Night at the Opera" track="01"
              Matches  the  first  track  on the album A Night at the Opera by
              Queen.  Note that track tags are interpreted  as  text  and  not
              numbers, meaning "01" is not the same as "1".

       artist="Clark \"Plazmataz\" Powell"
              Matches tracks by Clark "Plazmataz" Powell.

```

Kind regards,
Fabian

EDIT: updated rendered man page excerpt to reflect new file format

MaxKellermann commented on src/util/Compiler.h:
```diff
@@ -42,9 +42,11 @@
+#define unreachable() __assume(false)
 #else
-#define gcc_unreachable()
+#define unreachable()
```
Dropping this prefix will break the build with a C++23 compiler.

push0ret replied:
Oh, sorry about that. Will fix.

MaxKellermann commented on doc/mpdscribble.1:
```diff
@@ -126,6 +126,58 @@ Your Last.fm password, either cleartext or its MD5 sum.
 have a connection to the scrobbler.  This option used to be called
 "cache".  It is optional.
+.TP
+.B ignore = FILE
```
Do we really need an indirection with another file? This complicates the thing both for the code and for users.
Also this file format looks rather arbitrary and cannot be extended or changed later. One has to remember what those columns are.

push0ret replied:
I used a separate file for the following reasons:
- Avoid complicating the config file parser itself.
- Managing the lists is easier when they are separate files. Users may want to interact with these lists through other utilities, for example a shell script that gets the current song playing from mpd and writes the details into a list.
- Makes sharing the lists across multiple scrobblers simply work.

I came up with the file format to make the parser implementation trivial / easily verifiable. I wanted to avoid having weird audio metadata end up breaking the parser. I considered two options: Very simple easy to parse format and write a small parser myself, or use a standardized machine readable format and implement it using a library. I'm happy to discuss alternatives.
The format also allows easy construction of list entries from audio metadata. In most cases, a simple `mpc current -f "%artist%;%album%;%title%" >> ignorelist` should just work. For the rare cases the metadata itself contains a semicolon, the escaping rules are simple to apply via regex substitution.

push0ret replied:
Hi Max

I understand your concern about the inflexibility of the file format and to address this I would propose a "ignore_file_tags" (naming is not final) option that is global for all scrobblers.
This option would be set to a space delimited string of tag names, which will then correspond to the columns of the ignore file. The current behavior would be represented by the option `ignore_file_tags = artist album title`.

At the moment, the `Record` structure only exposes `artist`, `album`, `title` and `track` tags, so the flexibility is quite limited. However, making the format configurable allows us to add more tags in the future without breaking users existing configurations while also allowing users to only match the tags they care about in their ignore file, avoiding useless extra columns.

Please let me know if you have concerns with this approach. I will implement this after your green light.

MaxKellermann replied:
This idea sounds like it will grow into a complexity monster. And you're going to split the specification between the configuration file and the ignore file - the ignore file will no longer be self-contained, it depends on external data (the tag list specified in the configuration file).

push0ret replied:
Thanks for your feedback. 
I think based on this we could opt for one of the following options:
- Make the first line of the ignore file a mandatory format specification. E.g. `artist;album;title` as first line would set the format to what is hard coded now. I would avoid an optional format specification line as this would then require some marker to distinguish the specification line from an ignore line. Such a marker in turn requires an escaping mechanism to avoid clashing with metadata.
- Drop the configurable columns altogether. New tags can always be supported with additional columns at the end. This would not break existing configurations, as trailing semicolons are already not required. The current parser treats `Queen`, `Queen;` and `Queen;;` identically. Downside is potentially worse user experience if at some point many more tags are supported. Users may have to deal with an annoying number of empty columns in their ignore files.
- Scratch the custom format and parser, move to something standardized. I do like the minimal format we have at the moment, I think it's easy to deal with and doesn't pull in any dependencies. But if you prefer to use a standardized format, I can implement that (but will require a new dependency).

MaxKellermann replied:
Why not do something like:
```
ignore artist="Queen" album="A Night at the Opera"
ignore artist="Queen" title="Bohemian Rhapsody"
```
... right in the main configuration file. Simpler and super flexible.

push0ret replied:
I think the dedicated file is useful for:
- Automated management by tools/scripts (see my first comment).
- Making it easy to use an ignore list on a subset of scrobblers, without duplicating the lists themselves.

The reason I did not use the format you proposed in the ignore file is simply parser complexity, having to handle quoting properly. But that is obviously not a blocker.
Considering the aforementioned benefits of the dedicated file, would it be an acceptable compromise to keep the dedicated ignore file, but use a tag="value" format as you suggested?

With the cost of additional complexity, we could also support both methods (in-place in the main config and dedicated ignore files). I'm not positive that that's worth it, though.

MaxKellermann replied:
Quoting is difficult, of course - but your CSV-like syntax doesn't solve this - look at how you had to implement backslash-quoting. No difference here.

I think it's okay to have this external file. I don't quite like the idea of having two files, but it's a situation where all possible solutions are bad, and maybe you're right and it's the least-bad one.

push0ret replied:
Yeah, I agree. I will implement the tag="value" format for the ignore files then.
Thanks for taking the time to discuss this, it is appreciated.

MaxKellermann commented on src/Config.hxx:
```diff
@@ -17,7 +18,10 @@ NullableString(const std::string &s) noexcept
 
+
 struct Config {
+	using IgnoreListMap = std::map<std::string, std::shared_ptr<IgnoreList>>;
```
Why is this a `shared_ptr`? And if there really is a good reason to do this, why did you not include `<memory>`?

push0ret replied:
I agree refcounting is not needed.

I will switch the owning pointer to unique_ptr and the non-owning pointer in the ScrobblerConfig to a raw pointer. Please let me know if you disagree with this approach.

MaxKellermann replied:
Explain why you want the `unique_ptr` indirection

push0ret replied:
For automatic memory management. Do you propose to manually allocate and free?

MaxKellermann replied:
No, I don't propose to do that. Why would I want to do that? That would still be an indirection, just an unmanaged one. I don't want the indirection.

push0ret replied:
Oh right, sorry I misunderstood you. Indeed, the map value is stable so we can point to it directly. I will remove the indirection.

MaxKellermann commented on src/IgnoreList.cxx:
```diff
@@ -0,0 +1,40 @@
+#include <ranges>
+#include <algorithm>
+
+bool IgnoreListEntry::matches_record(const Record& record) const {
```
Bad coding style, and you forgot `noexcept`

push0ret replied:
Apologies, will fix.

MaxKellermann commented on src/IgnoreList.cxx:
```diff
@@ -0,0 +1,40 @@
+bool IgnoreListEntry::matches_record(const Record& record) const {
+	// The below logic would always return true if the entry is empty.
+	// This condition should never be true, as we don't push empty entries.
+	assert(this->artist || this->album || this->title);
```
Please don't use `this->` everywhere, it's redundant noise.

push0ret replied:
Will fix.

MaxKellermann commented on src/IgnoreList.cxx:
```diff
@@ -0,0 +1,40 @@
+#include <algorithm>
+
+bool IgnoreListEntry::matches_record(const Record& record) const {
+	// The below logic would always return true if the entry is empty.
+	// This condition should never be true, as we don't push empty entries.
```
Did you know that there's a special syntax for multi-line comments, and that `//` is the syntax for explicitly single-line comments?

push0ret replied:
Apologies.

MaxKellermann commented on src/IgnoreList.cxx:
```diff
@@ -0,0 +1,40 @@
+	// This condition should never be true, as we don't push empty entries.
+	assert(this->artist || this->album || this->title);
+
+	if (this->artist) {
+		if (record.artist.empty() || record.artist != this->artist) {
+			return false;
+		}
+	}
+
+	if (this->album) {
+		if (record.album.empty() || record.album != this->album) {
+			return false;
+		}
+	}
+
+	if (this->title) {
+		if (record.track.empty() || record.track != this->title) {
+			return false;
+		}
+	}
```
That's 3 copies of the same piece of code

push0ret replied:
Hm, would you like me to reduce the boilerplate with a macro? I don't see a better way right now.

MaxKellermann replied:
```
[[gnu::pure]]
static constexpr bool
IsIgnoreMismatch(std::string_view ignore, std::string_view value) noexcept
{
  return !ignore.empty() && (value.empty() || value != ignore);
}
```
Simple, eh?

push0ret replied:
Yeah, true. Will fix this.

MaxKellermann commented on src/IgnoreList.hxx:
```diff
@@ -0,0 +1,28 @@
+#include "Record.hxx"
+
+struct IgnoreListEntry {
+	std::optional<std::string> artist;
```
The config file uses an empty string for "match everything". Why wrap this in `std::optional` here? All this does is add overhead for nothing.

push0ret replied:
That's true. I added optional only to express the semantics in the code. Happy to reduce the overhead though and just go with the strings. Will fix.

MaxKellermann commented on src/ReadConfig.cxx:
```diff
@@ -21,6 +21,9 @@
 
 #include <stdlib.h>
 #include <string.h>
+#include <filesystem>
```
`<filesystem>` shall not be used

MaxKellermann commented on src/ReadConfig.cxx:
```diff
@@ -21,6 +21,9 @@
 #include <stdlib.h>
 #include <string.h>
+#include <filesystem>
+#include <fstream>
```
`<iostreams>` shall not be used

push0ret replied:
Okay, I will use C style file handling then.

MaxKellermann commented on src/ReadConfig.cxx:
```diff
@@ -231,10 +234,128 @@ load_unsigned(const IniFile &file, const char *name, unsigned *value_r)
+
+			IgnoreListEntry entry{};
+
+			// Format:
+			//   artist;album;title
+			// Empty string means "ignore", so e.g.:
+			//   ;;title               Match any track titled "title".
+			//   artist                Match any track with artist "artist".
+			//   artist;;              Identical to the above.
+			//   artist;;title         Match any track titled "title" from artist "artist".
+			//   artist;album;title    Match everything.
+			//   ;;                    Ignored, empty entries will not be pushed.
+			//
+			// The format is designed easy to parse and unlikely to conflict with the metadata
+			// we see in the wild. Realistically, the only conflict you may get is a semicolon in
+			// the metadata, which you can escape with a single backslash, e.g.:
+			//   artist1\;artist2;;   Match artist "artist1;artist2"
+			//
+			// Note that the parser does not have entirely typical support for escaping:
+			// Escaping rules are only applied for the sequence immediately before a semicolon.
+			// Backslashes do not have any special meaning outside this context, so double backslashes
+			// are not collapsed anywhere else.
+			// When interpreted as an escape, backslashes are collapsed as you would expect.
+			//   artist1\\artist2;album         Match "artist1\\artist2" and album "album"
+			//   artist\\\\;album;              Match artist "artist\\" and album "album"
+			//   artist1\\\\\;artist2;album;    Match artist "artist1\\;artist2" and album "album"
```
Sigh. Yet another huge multi-line comment with the single-line comment syntax

MaxKellermann commented on src/ReadConfig.cxx:
```diff
@@ -231,10 +234,128 @@ load_unsigned(const IniFile &file, const char *name, unsigned *value_r)
+					throw FormatRuntimeError("Invalid ignore specification (too many tokens), line: %d", line_num);
+				}
+
+				std::string token = temp_token + std::string(part.begin(), part.end());
```
The temporary `std::string` constructed here is pointless overhead

push0ret replied:
Right, will fix.

MaxKellermann commented:
Your way of splitting the PR into commits doesn't make sense. If you have a known-bad commit, don't add another commit to fix it; instead, amend the bad commit.

About `UNREACHABLE`: let's just not rename it. Yes, the "gcc_" prefix isn't exactly correct, but making all UPPERCASE is a different kind of ugliness, but doesn't improve anything. This whole macro is obsolete anyway - once we have C++23, we can simply do `std::unreachable()` and remove the macro completely.

push0ret commented:
> Your way of splitting the PR into commits doesn't make sense. If you have a known-bad commit, don't add another commit to fix it; instead, amend the bad commit.

Okay, then I will for the most part squash this down with the next patch.

> About UNREACHABLE: let's just not rename it.

Okay, I'll revert to the original name.

push0ret commented:
Hi Max

I updated the file format as discussed previously and also implemented the other requested changes.
The  changes are squashed into the original two commits.

MaxKellermann commented:
```
../../src/ReadConfig.cxx: In function ‘IgnoreList* load_ignore_list(const std::string&, Config::IgnoreListMap&)’:
../../src/ReadConfig.cxx:322:14: error: ‘file_exists’ was not declared in this scope; did you mean ‘std::errc::file_exists’?
  322 |         if (!file_exists(path.c_str())) {
      |              ^~~~~~~~~~~
      |              std::errc::file_exists
```
Windows build fails. And I wonder why this check exists at all - what is the point?

push0ret commented:
> ```
> ../../src/ReadConfig.cxx: In function ‘IgnoreList* load_ignore_list(const std::string&, Config::IgnoreListMap&)’:
> ../../src/ReadConfig.cxx:322:14: error: ‘file_exists’ was not declared in this scope; did you mean ‘std::errc::file_exists’?
>   322 |         if (!file_exists(path.c_str())) {
>       |              ^~~~~~~~~~~
>       |              std::errc::file_exists
> ```
> 
> Windows build fails. And I wonder why this check exists at all - what is the point?

Hi Max!

Sorry for breaking the windows build there, I missed the preprocessor conditional around the `file_exists()` implementation.

The check is intended to provide more accurate error messages, differentiating between file does not exist and file cannot be opened (typically due to insufficient permissions). It's not strictly necessary and I also missed that `file_exists()` actually checks for file is regular, so it anyway does not behave exactly as I intended.

As I'm not going to implement a Windows API based `file_exists()` I will just drop the additional check.
